ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/TransactionLogUpdatesConversions.scala

```diff
@@ -0,0 +1,423 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.appendonlydao.events
+
+import java.time.Instant
+
+import com.daml.ledger.api.v1.transaction.{
+  TransactionTree,
+  TreeEvent,
+  Transaction => FlatTransaction,
+}
+import com.daml.ledger.api.v1.{event => apiEvent}
+import com.daml.lf.data.Ref
+import com.daml.logging.LoggingContext
+import com.daml.platform.ApiOffset
+import com.daml.platform.api.v1.event.EventOps.TreeEventOps
+import com.daml.platform.participant.util.LfEngineToApi
+import com.daml.platform.store.interfaces.TransactionLogUpdate
+import com.daml.platform.store.interfaces.TransactionLogUpdate.{CreatedEvent, ExercisedEvent}
+import com.google.protobuf.timestamp.Timestamp
+
+import scala.collection.mutable
+import scala.concurrent.{ExecutionContext, Future}
+
+private[events] object TransactionLogUpdatesConversions {
+  object ToFlatTransaction {
+    def apply(
+        tx: TransactionLogUpdate.Transaction,
+        filter: FilterRelation,
+        verbose: Boolean,
+        lfValueTranslation: LfValueTranslation,
+    )(implicit
+        loggingContext: LoggingContext,
+        executionContext: ExecutionContext,
+    ): Future[Option[FlatTransaction]] = {
+      val transactionEvents = tx.events
+      val filtered = transactionEvents.filter(FlatTransactionPredicate(_, filter))
+
+      filtered.headOption
+        .map { first =>
+          val events = removeTransient(filtered)
+
+          Future
```

Comment:
We are spawning here a number of futures on behalf of a single transaction. Isn't that too granular concurency?

Reply:
It is more granular than before, but I didn't consider it a problem. We are wrapping some cpu significant ops in these Futures, so I wouldn't expect a considerable impact. I'll play with some benchmarks just to be sure

Reply:
Let's test it and revisit if and when it causes issues